File: java_solutions/2563.count-the-number-of-fair-pairs.java
/*
 * @lc app=leetcode id=2563 lang=java
 *
 * [2563] Count the Number of Fair Pairs
 */

class Solution {
    public long countFairPairs(int[] nums, int lower, int upper) {
        return Stream.of(IntStream.range(0, nums.length).mapToLong(i -> (long) nums[i] << 20 | i).sorted().toArray())
            .mapToLong(A -> IntStream.range(0, A.length)
                .mapToLong(i -> (long) (~Arrays.binarySearch(A, i + 1, A.length, (long) (upper - (A[i] >> 20)) << 20 | 1048575L)) 
                                     - (~Arrays.binarySearch(A, i + 1, A.length, (long) (lower - (A[i] >> 20) - 1) << 20 | 1048575L)))
                .sum())
            .findFirst().getAsLong();
    }
}

File: kotlin_solutions/2563.count-the-number-of-fair-pairs.kt
/*
 * @lc app=leetcode id=2563 lang=kotlin
 *
 * [2563] Count the Number of Fair Pairs
 */

class Solution {fun countFairPairs(nums: IntArray, lower: Int, upper: Int) = if (nums.size == 1) 0L else nums.apply { sort() }.indices.fold(0L to (TreeMap<Int, Int>() to TreeMap<Int, Int>())) { (count, maps), i -> maps.let { (map1, map2) -> map1.apply { putIfAbsent(nums[i], i) } to map2.apply { put(nums[i], i) } }.let { (map1, map2) -> (count + (map2.ceilingKey(lower - nums[i])?.let { l -> map2.floorKey(upper - nums[i])?.takeIf { r -> l <= r }?.let { r -> map2[r]!! - map1[l]!! + if (map2[r] != i) 1 else 0 } } ?: 0)) to (map1 to map2) } }.first}
